drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,756 @@
+{
+	struct net_buf *buf;
+	int err;
+
+	struct usb_setup_packet *setup_packet =
+		(struct usb_setup_packet *)&evt->hal_evt.setup_received;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return -ENOMEM;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup_packet, sizeof(struct usb_setup_packet));
+	net_buf_add(buf, sizeof(struct usb_setup_packet));
+
+	/* Update to next stage of control transfer */
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		LOG_DBG("s:%p|feed for -out-", buf);
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			err = udc_submit_ep_event(dev, buf, err);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		err = udc_ctrl_submit_s_in_status(dev);
+	} else {
+		err = udc_ctrl_submit_s_status(dev);
+	}
+
+	return err;
+}
+
+static void udc_event_xfer_ctrl_in(const struct device *dev, struct net_buf *const buf)
+{
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+	if (udc_ctrl_stage_is_status_in(dev) || udc_ctrl_stage_is_no_data(dev)) {
+		/* Status stage finished, notify upper layer */
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_status_out(dev)) {
+		/* Status stage finished, notify upper layer */
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	/* Clear setup endpoint. Send a ZLP */
+	R_USBD_XferStart(&data->udc, USB_CONTROL_EP_IN, NULL, 0);
+}
+
+static void udc_event_status_in(const struct device *dev)
+{
+	struct net_buf *buf;
+
+	buf = udc_buf_get(dev, USB_CONTROL_EP_IN);
+	if (unlikely(buf == NULL)) {
+		LOG_DBG("ep 0x%02x queue is empty", USB_CONTROL_EP_IN);
+		return;
+	}
+
+	udc_event_xfer_ctrl_in(dev, buf);
+}
+
+static void udc_event_xfer_ctrl_out(const struct device *dev, struct net_buf *const buf,
+				    uint32_t len)
+{
+	net_buf_add(buf, len);
+
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_status_out(dev)) {
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	if (udc_ctrl_stage_is_status_in(dev)) {
+		udc_ctrl_submit_s_out_status(dev, buf);
+	}
+}
+
+static void udc_event_xfer_out(const struct device *dev, struct net_buf *const buf, uint32_t len)
+{
+	net_buf_add(buf, len);
+	udc_submit_ep_event(dev, buf, 0);
+}
+
+static void udc_event_xfer_in(const struct device *dev, struct net_buf *const buf)
+{
+	udc_submit_ep_event(dev, buf, 0);
+}
+
+static void udc_event_xfer_complete(const struct device *dev, struct udc_renesas_ra_evt *evt)
+{
+	struct net_buf *buf;
+
+	unsigned int ep = evt->hal_evt.xfer_complete.ep_addr;
+	unsigned int result = evt->hal_evt.xfer_complete.result;
+	unsigned int len = evt->hal_evt.xfer_complete.len;
+
+	buf = udc_buf_get(dev, ep);
+
+	if (buf == NULL) {
+		return;
+	}
+
+	if (result == USBD_XFER_RESULT_SUCCESS) {
+		udc_ep_set_busy(dev, ep, 0);
```

Comment:
Comment from @jfischer-no about ZLP could be potentially solved here
```c
		/* ZLP needed? */
		if (USB_EP_DIR_IS_IN(ep) && udc_ep_buf_has_zlp(buf)) {
			udc_ep_buf_clear_zlp(buf);
			/* Send ZLP, notification about transfer complete should come again */
			... = R_USBD_XferStart(udc, USB_CONTROL_EP_IN, NULL, 0);
			return;
		}
```

To test that it works, code could make some string descriptor that will have 64 bytes total. Maybe manufacturer name or serial number. This case then could be tested during enumeration.

**But udc_buf_get() should be change to udc_buf_peek() for ZLP !!!**

Reply:
I fixed it

Reply:
Control IN transfers do have ZLP during Data Stage only if the payload is multiple of `bMaxPacketSize0` and the host requested `wLength` greater than multiple of `bMaxPacketSize0`.